polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -946,44 +937,49 @@ impl<T: Config> PreCodeUpgrade for Pallet<T> {
 
 		let current_deposit = info.deposit;
 
+		let lease_holding = Self::is_parachain(para);
 		let free_upgrade = requirements == UpgradeRequirements::SkipRequirements ||
 			para.is_system() ||
 			lease_holding;
 
-		match (free_upgrade, info.billing_account.clone()) {
-			(false, Some(billing_account)) => {
-				if let Err(_) = <T as Config>::Currency::withdraw(
-					&billing_account,
-					T::UpgradeFee::get(),
-					WithdrawReasons::FEE,
-					ExistenceRequirement::KeepAlive,
-				) {
-					Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
-						CodeUpgradeScheduleError::FailedToPayUpgradeFee,
-					));
-					// This is an overestimate of the used weight, but it's better to be safe than
-					// sorry.
-					return Err(<T as Config>::WeightInfo::pre_code_upgrade())
-				}
+		if !free_upgrade {
+			if info.billing_account.is_none() {
+				Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
+					CodeUpgradeScheduleError::BillingAccountNotSet,
+				));
+				// An overestimate of the used weight, but it's better to be safe than sorry.
+				return Err(<T as Config>::WeightInfo::pre_code_upgrade())
+			}
 
-				let additional_deposit = new_deposit.saturating_sub(current_deposit);
-				if let Err(_) =
-					<T as Config>::Currency::reserve(&billing_account, additional_deposit)
-				{
-					Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
-						CodeUpgradeScheduleError::FailedToReserveDeposit,
-					));
-					// This is an overestimate of the used weight, but it's better to be safe than
-					// sorry.
-					return Err(<T as Config>::WeightInfo::pre_code_upgrade())
-				}
+			let billing_account = info
+				.billing_account
+				.clone()
+				.expect("Ensured above that the billing account is set; qed");
```

Comment:
The previous pattern match was good. Checking with `is_none` and then expecting is unnecessarily brittle.

Reply:
E.g. 

```rust
let Some(billing_account) = info.billing_account else {
  Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
	CodeUpgradeScheduleError::BillingAccountNotSet,
	
  ));
  // An overestimate of the used weight, but it's better to be safe than sorry.
  return Err(<T as Config>::WeightInfo::pre_code_upgrade())
}

```